File: java/playwright/src/main/java/web_playwright/WebFormPage.java
package web_playwright;

import java.nio.file.Paths;

import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.AriaRole;
import com.microsoft.playwright.Locator;

public class WebFormPage {

    private Page page;
    private Locator input;
    private Locator password;
    private Locator textarea;
    private Locator number;
    private Locator city;
    private Locator file;
    private Locator color;
    private Locator date;
    private Locator range;
    private Locator button;

    public WebFormPage(Page page) {
        this.page = page;
        this.input = page.getByLabel("Text input");
        this.password = page.getByLabel("Password");
        this.textarea = page.getByLabel("Textarea");
        this.number = page.locator("//select[@name='my-select']");
        this.city = page.locator("//input[@name='my-datalist']");
        this.color = page.getByLabel("Color picker");
        this.date = page.getByLabel("Date picker");
        this.range = page.getByLabel("Example range");
        this.file = page.getByLabel("File input");
        this.button = page.getByRole(AriaRole.BUTTON, new Page.GetByRoleOptions().setName("Submit"));
    }

    public void set_input(String value) {
        this.input.fill(value);
    }

    public void set_password(String value) {
        this.password.fill(value);
    }

    public void set_textarea(String value) {
        this.textarea.fill(value);
    }

    public void set_number(int value) {
        this.number.selectOption(String.valueOf(value));
    }

    public void set_city(String value) {
        this.city.fill(value);
    }

    public void set_file(String value) {
        this.file.setInputFiles(Paths.get(value));
    }

    public void set_color(String value) {
        this.color.fill(value);
    }

    public void set_date(String value) {
        this.date.evaluate("(node, val) => { node.value = val; }", value);
    }

    public void set_range(int value) {
        this.range.fill(String.valueOf(value));
    }

    public void submit() {
        this.button.click();
    }

}

File: python/playwright/pages/webform_page.py
from playwright.sync_api import Page, expect


class WebFormPage:

    def __init__(self, page: Page):
        self.page = page
        self.input = page.locator("[id='my-text-id']")
        self.password = page.locator("[name='my-password']")
        self.textarea = page.locator("[name='my-textarea']")
        self.number = page.locator("//select[@name='my-select']")
        self.city = page.locator("//input[@name='my-datalist']")
        self.color = page.locator("[name='my-colors']")
        self.date = page.locator("[name='my-date']")
        self.range = page.locator("[name='my-range']")
        self.file = page.locator("[name='my-file']")
        self.button_submit = page.get_by_role("button", name="Submit")

    def set_input(self, value):
        self.input.fill(value)

    def set_password(self, value):
        self.password.fill(value)

    def set_textarea(self, value):
        self.textarea.fill(value)

    def set_number(self, value):
        self.number.select_option(index=value)

    def set_city(self, value):
        self.city.fill(value)

    def set_color(self, value):
        self.color.fill(value)

    def set_date(self, value):
        #self.date.fill(value)
        # Click elsewhere to close the calendar
        #self.page.locator("body").click()
        self.date.evaluate("(elem, val) => elem.setAttribute('value', val);", value)

    def set_range(self, value):
        # Two solutions
        # Solution #1: Trigger mouse events to move the range handler
        width = self.range.evaluate("elem => { return elem.getBoundingClientRect().width }")
        self.range.hover(position={'x': 0, 'y': 0})
        self.page.mouse.down()
        self.range.hover(position={'x': (width * value) / 10, 'y': 0})
        self.page.mouse.up()

        # solution #2: Set the range value directly
        self.range.fill(str(value))

    def set_file(self, value):
        self.file.set_input_files(value)

    def submit(self):
        self.button_submit.click()
